### ISSUE TITLE
NNPack support for arm ( linux )

### ISSUE THREAD

**Issue by danFromTelAviv** [pytorch/pytorch#17714]
## 🚀 Feature
allow nnpack to install gracefully with pytorch on linux arm devices 

## Motivation

pytorch+NNPack should be one of the fastest options for arm. 

## Pitch
pytorch is already pretty easy to install on arm. Together with nnpack pytorch will be posed very well in the competition of iot deep learning. The delta is relatively small 👍 The spoils are large.
Bullet format:
1) NNpack easily installs with ninja on arm [tutorial for darnknet + NNpack](http://funofdiy.blogspot.com/2018/08/deep-learning-with-raspberry-pi-real.html)
2) Pytorch is the shit for developing deep learning. 
3) [Pytorch already installs easily without nnpack.](https://medium.com/hardware-interfacing/how-to-install-pytorch-v4-0-on-raspberry-pi-3b-odroids-and-other-arm-based-devices-91d62f2933c7) 
4) arm IOT is a big chunck of the future of deep learning. 
5) Pytorch can be the shit for deploying deep learning on arm devices.

It's a great low hanging fruit thats very good for PR. "pytorch is the fastest deep learning framework on the raspberry pi".

*** please please please ***
you can easily beat tf here.... just saying.

**Comment by t-vi** [comment]
PyTorch master should have NNPack support all right and uses it for convolutions (this was https://github.com/pytorch/pytorch/pull/15924).
Could you be more specific in what you find lacking, I'd be glad to improve things.

**Comment by danFromTelAviv** [comment]
```
export USE_NNPACK=1 
python3 setup.py build 
```
on arm linux fails. 
( this is also true of qnnpack - which would also be nice to have ). 

There is a caffe2 build .sh file but it also has other issues.

**Comment by t-vi** [comment]
I don't have an arm development environment other than Android, so I can't comment. Would you share the error message?

**Comment by soumith** [comment]
@danFromTelAviv it works for me, so there must be something weird about your setup. Please paste full build log.

**Comment by danFromTelAviv** [comment]
It may be worth mentioning that I'm working on a tinkerboard ( very similar to a raspberry pi but better performance  -mfpu=neon-vfpv4 -mtune=cortex-a17 -march=armv7ve -mfloat-abi=hard) and ubuntu 18.04 - not raspbian ( although both are debian based ). 
@soumith  - I saw in blogs online that people with raspberry pi ( which I assume worked for you ) could not install pytorch with nnpack and qnnpack ( and cuda and distributed obviously ).  
Can you specify what exactly worked for you and on what platform / architecture / os. 
I'll run it again today and print out the errors ( it should take a few hours before it gets to the errors ). 

Thanks both for your quick responses!

**Comment by danFromTelAviv** [comment]
while its running this is the procedure i'm running:

`
sudo apt update && sudo apt upgrade
`
`
sudo apt install libopenblas-dev libblas-dev m4 cmake cython python3-dev python3-yaml python3-setuptools
`
`cd ~`
`mkdir pytorch_install && cd pytorch_install`
`git clone --recursive https://github.com/pytorch/pytorch`
`cd pytorch`

`export MAX_JOBS=2`
`export NO_CUDA=1`
`export NO_DISTRIBUTED=1`
`export NO_MKLDNN=1`
`export USE_NNPACK=1 `
`export USE_QNNPACK=1 `
`export USE_CFLAGS=1`
`export CFLAGS=’-mfpu=neon-vfpv4 -mtune=cortex-a17 -march=armv7ve -mfloat-abi=hard’`

`python3 setup.py build # clean first if needed`
`sudo -E python3 setup.py install`

**Comment by danFromTelAviv** [comment]
so a twist in the plot : 

```
-- Building version 1.1.0a0+1d827b7
PyTorch is only supported to run with 64-bit python.
```

**Comment by danFromTelAviv** [comment]
it begins:
> root@localhost:~/pytorch_install/pytorch# python3 setup.py build
> Building wheel torch-1.1.0a0+1d827b7
> -- Building version 1.1.0a0+1d827b7
> ['cmake',
>  '/root/pytorch_install/pytorch',
>  '-DBUILDING_WITH_TORCH_LIBS=ON',
>  '-DBUILD_BINARY=False',
>  '-DBUILD_CAFFE2_OPS=True',
>  '-DBUILD_PYTHON=True',
>  '-DBUILD_SHARED_LIBS=ON',
>  '-DBUILD_TEST=True',
>  '-DBUILD_TORCH=ON',
>  '-DCAFFE2_STATIC_LINK_CUDA=False',
>  '-DCMAKE_BUILD_TYPE=Release',
>  '-DCMAKE_CXX_FLAGS=-mfpu=neon-vfpv4 -mtune=cortex-a17 -march=armv7ve '
>  '-mfloat-abi=hard ',
>  '-DCMAKE_C_FLAGS=-mfpu=neon-vfpv4 -mtune=cortex-a17 -march=armv7ve '
>  '-mfloat-abi=hard ',
>  '-DCMAKE_EXE_LINKER_FLAGS=',
>  '-DCMAKE_INSTALL_PREFIX=/root/pytorch_install/pytorch/torch',
>  '-DCMAKE_PREFIX_PATH=/usr/lib/python3/dist-packages',
>  '-DCMAKE_SHARED_LINKER_FLAGS=',
>  '-DINSTALL_TEST=True',
>  '-DNCCL_EXTERNAL=False',
>  '-DNUMPY_INCLUDE_DIR=/usr/local/lib/python3.5/dist-packages/numpy/core/include',
>  '-DONNX_ML=False',
>  '-DONNX_NAMESPACE=onnx_torch',
>  '-DPYTHON_EXECUTABLE=/usr/bin/python3',
>  '-DPYTHON_INCLUDE_DIR=/usr/include/python3.5m',
>  '-DPYTHON_LIBRARY=/usr/lib/libpython3.5m.so.1.0',
>  '-DTHD_SO_VERSION=1',
>  '-DTORCH_BUILD_VERSION=1.1.0a0+1d827b7',
>  '-DUSE_CUDA=False',
>  '-DUSE_DISTRIBUTED=False',
>  '-DUSE_FBGEMM=True',
>  '-DUSE_FFMPEG=False',
>  '-DUSE_LEVELDB=False',
>  '-DUSE_LMDB=False',
>  '-DUSE_MKLDNN=False',
>  '-DUSE_NCCL=False',
>  '-DUSE_NNPACK=True',
>  '-DUSE_NUMPY=True',
>  '-DUSE_OPENCV=False',
>  '-DUSE_QNNPACK=True',
>  '-DUSE_ROCM=False',
>  '-DUSE_SYSTEM_EIGEN_INSTALL=OFF',
>  '-DUSE_SYSTEM_NCCL=False',
>  '-DUSE_TENSORRT=False']
> -- The CXX compiler identification is GNU 5.3.1
> -- The C compiler identification is GNU 5.3.1
> -- Check for working CXX compiler: /usr/bin/c++
> -- Check for working CXX compiler: /usr/bin/c++ -- works
> -- Detecting CXX compiler ABI info
> -- Detecting CXX compiler ABI info - done
> -- Detecting CXX compile features
> -- Detecting CXX compile features - done
> -- Check for working C compiler: /usr/bin/cc
> -- Check for working C compiler: /usr/bin/cc -- works
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - done
> -- Detecting C compile features
> -- Detecting C compile features - done
> -- Not forcing any particular BLAS to be found
> -- Performing Test COMPILER_WORKS
> -- Performing Test COMPILER_WORKS - Success
> -- Performing Test SUPPORT_GLIBCXX_USE_C99
> -- Performing Test SUPPORT_GLIBCXX_USE_C99 - Success
> -- Performing Test CAFFE2_EXCEPTION_PTR_SUPPORTED
> -- Performing Test CAFFE2_EXCEPTION_PTR_SUPPORTED - Success
> -- std::exception_ptr is supported.
> -- Performing Test CAFFE2_IS_NUMA_AVAILABLE
> -- Performing Test CAFFE2_IS_NUMA_AVAILABLE - Failed
> -- NUMA is not available
> -- Performing Test CAFFE2_NEED_TO_TURN_OFF_DEPRECATION_WARNING
> -- Performing Test CAFFE2_NEED_TO_TURN_OFF_DEPRECATION_WARNING - Success
> -- Performing Test CAFFE2_COMPILER_SUPPORTS_AVX2_EXTENSIONS
> -- Performing Test CAFFE2_COMPILER_SUPPORTS_AVX2_EXTENSIONS - Failed
> -- Performing Test CAFFE2_COMPILER_SUPPORTS_AVX512_EXTENSIONS
> -- Performing Test CAFFE2_COMPILER_SUPPORTS_AVX512_EXTENSIONS - Failed
> -- Performing Test COMPILER_SUPPORTS_HIDDEN_VISIBILITY
> -- Performing Test COMPILER_SUPPORTS_HIDDEN_VISIBILITY - Success
> -- Performing Test COMPILER_SUPPORTS_HIDDEN_INLINE_VISIBILITY
> -- Performing Test COMPILER_SUPPORTS_HIDDEN_INLINE_VISIBILITY - Success
> -- Performing Test COMPILER_SUPPORTS_RDYNAMIC
> -- Performing Test COMPILER_SUPPORTS_RDYNAMIC - Success
> -- Building using own protobuf under third_party per request.
> -- Use custom protobuf build.
> -- Looking for pthread.h
> -- Looking for pthread.h - found
> -- Looking for pthread_create
> -- Looking for pthread_create - not found
> -- Looking for pthread_create in pthreads
> -- Looking for pthread_create in pthreads - not found
> -- Looking for pthread_create in pthread
> -- Looking for pthread_create in pthread - found
> -- Found Threads: TRUE
> -- Caffe2 protobuf include directory: $<BUILD_INTERFACE:/root/pytorch_install/pytorch/third_party/protobuf/src>$<INSTALL_INTERFACE:include>
> -- Trying to find preferred BLAS backend of choice: MKL
> -- Looking for sys/types.h
> -- Looking for sys/types.h - found
> -- Looking for stdint.h
> -- Looking for stdint.h - found
> -- Looking for stddef.h
> -- Looking for stddef.h - found
> -- Check size of void*
> -- Check size of void* - done
> CMake Warning at cmake/Dependencies.cmake:118 (message):
>   MKL could not be found.  Defaulting to Eigen
> Call Stack (most recent call first):
>   CMakeLists.txt:224 (include)
> 
> 
> CMake Warning at cmake/Dependencies.cmake:137 (message):
>   Preferred BLAS (MKL) cannot be found, now searching for a general BLAS
>   library
> Call Stack (most recent call first):
>   CMakeLists.txt:224 (include)
> 
> 
> -- Checking for [mkl_intel - mkl_gnu_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_intel - mkl_intel_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf - mkl_gnu_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_gf - mkl_intel_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel - mkl_gnu_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_intel - mkl_intel_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf - mkl_gnu_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_gf - mkl_intel_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel - mkl_gnu_thread - mkl_core - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_intel - mkl_intel_thread - mkl_core - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf - mkl_gnu_thread - mkl_core - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_gf - mkl_intel_thread - mkl_core - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel - mkl_sequential - mkl_core - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf - mkl_sequential - mkl_core - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel - mkl_core - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf - mkl_core - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl - guide - pthread - m]
> --   Library mkl: not found
> -- MKL library not found
> -- Checking for [Accelerate]
> --   Library Accelerate: BLAS_Accelerate_LIBRARY-NOTFOUND
> -- Checking for [vecLib]
> --   Library vecLib: BLAS_vecLib_LIBRARY-NOTFOUND
> -- Checking for [openblas]
> --   Library openblas: /usr/lib/libopenblas.so
> -- Looking for sgemm_
> -- Looking for sgemm_ - found
> -- Performing Test BLAS_F2C_DOUBLE_WORKS
> -- Performing Test BLAS_F2C_DOUBLE_WORKS - Failed
> -- Performing Test BLAS_F2C_FLOAT_WORKS
> -- Performing Test BLAS_F2C_FLOAT_WORKS - Success
> -- Performing Test BLAS_USE_CBLAS_DOT
> -- Performing Test BLAS_USE_CBLAS_DOT - Success
> -- Found a library with BLAS API (open).
> -- The ASM compiler identification is GNU
> -- Found assembler: /usr/bin/cc
> -- Check if compiler accepts -pthread
> -- Check if compiler accepts -pthread - yes
> -- Brace yourself, we are building NNPACK
> -- Performing Test COMPILER_SUPPORTS_FP16_FORMAT
> -- Performing Test COMPILER_SUPPORTS_FP16_FORMAT - Success
> -- Found PythonInterp: /usr/bin/python3 (found version "3.5.1")
> -- Found Git: /usr/bin/git (found version "2.7.4")
> -- git Version: v1.4.0-505be96a
> -- Version: 1.4.0
> -- Performing Test HAVE_CXX_FLAG_STD_CXX11
> -- Performing Test HAVE_CXX_FLAG_STD_CXX11 - Success
> -- Performing Test HAVE_CXX_FLAG_WALL
> -- Performing Test HAVE_CXX_FLAG_WALL - Success
> -- Performing Test HAVE_CXX_FLAG_WEXTRA
> -- Performing Test HAVE_CXX_FLAG_WEXTRA - Success
> -- Performing Test HAVE_CXX_FLAG_WSHADOW
> -- Performing Test HAVE_CXX_FLAG_WSHADOW - Success
> -- Performing Test HAVE_CXX_FLAG_WERROR
> -- Performing Test HAVE_CXX_FLAG_WERROR - Success
> -- Performing Test HAVE_CXX_FLAG_PEDANTIC
> -- Performing Test HAVE_CXX_FLAG_PEDANTIC - Success
> -- Performing Test HAVE_CXX_FLAG_PEDANTIC_ERRORS
> -- Performing Test HAVE_CXX_FLAG_PEDANTIC_ERRORS - Success
> -- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32
> -- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32 - Failed
> -- Performing Test HAVE_CXX_FLAG_WFLOAT_EQUAL
> -- Performing Test HAVE_CXX_FLAG_WFLOAT_EQUAL - Success
> -- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING
> -- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING - Success
> -- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED_DECLARATIONS
> -- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED_DECLARATIONS - Success
> -- Performing Test HAVE_CXX_FLAG_WSTRICT_ALIASING
> -- Performing Test HAVE_CXX_FLAG_WSTRICT_ALIASING - Success
> -- Performing Test HAVE_CXX_FLAG_WD654
> -- Performing Test HAVE_CXX_FLAG_WD654 - Failed
> -- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY
> -- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY - Failed
> -- Performing Test HAVE_CXX_FLAG_COVERAGE
> -- Performing Test HAVE_CXX_FLAG_COVERAGE - Success
> -- Performing Test HAVE_STD_REGEX
> -- Performing Test HAVE_STD_REGEX
> -- Performing Test HAVE_STD_REGEX -- success
> -- Performing Test HAVE_GNU_POSIX_REGEX
> -- Performing Test HAVE_GNU_POSIX_REGEX
> -- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
> -- Performing Test HAVE_POSIX_REGEX
> -- Performing Test HAVE_POSIX_REGEX
> -- Performing Test HAVE_POSIX_REGEX -- success
> -- Performing Test HAVE_STEADY_CLOCK
> -- Performing Test HAVE_STEADY_CLOCK
> -- Performing Test HAVE_STEADY_CLOCK -- success
> CMake Warning at cmake/Dependencies.cmake:362 (message):
>   A compiler with AVX512 support is required for FBGEMM.  Not compiling with
>   FBGEMM.  Turn this warning off by USE_FBGEMM=OFF.
> Call Stack (most recent call first):
>   CMakeLists.txt:224 (include)
> 
> 
> -- Could NOT find Numa (missing:  Numa_INCLUDE_DIR Numa_LIBRARIES)
> CMake Warning at cmake/Dependencies.cmake:443 (message):
>   Not compiling with NUMA.  Suppress this warning with -DUSE_NUMA=OFF
> Call Stack (most recent call first):
>   CMakeLists.txt:224 (include)
> 
> 
> -- Using third party subdirectory Eigen.
> Python 3.5.1+
> -- Found PythonInterp: /usr/bin/python3 (found suitable version "3.5.1", minimum required is "2.7")
> -- Found PythonLibs: /usr/lib/libpython3.5m.so.1.0 (found suitable version "3.5.1+", minimum required is "2.7")
> CMake Warning at cmake/Dependencies.cmake:630 (find_package):
>   Could not find a package configuration file provided by "pybind11" with any
>   of the following names:
> 
>     pybind11Config.cmake
>     pybind11-config.cmake
> 
>   Add the installation prefix of "pybind11" to CMAKE_PREFIX_PATH or set
>   "pybind11_DIR" to a directory containing one of the above files.  If
>   "pybind11" provides a separate development package or SDK, be sure it has
>   been installed.
> Call Stack (most recent call first):
>   CMakeLists.txt:224 (include)
> 
> 
> -- Could NOT find pybind11 (missing:  pybind11_INCLUDE_DIR)
> -- Using third_party/pybind11.
> -- Adding OpenMP CXX_FLAGS: -fopenmp
> -- Will link against OpenMP libraries: /usr/lib/gcc/arm-linux-gnueabihf/5/libgomp.so;/usr/lib/arm-linux-gnueabihf/libpthread.so
> CMake Warning at cmake/Dependencies.cmake:994 (message):
>   Metal is only used in ios builds.
> Call Stack (most recent call first):
>   CMakeLists.txt:224 (include)
> 
> 
> --
> -- ******** Summary ********
> --   CMake version         : 3.5.1
> --   CMake command         : /usr/bin/cmake
> --   System                : Linux
> --   C++ compiler          : /usr/bin/c++
> --   C++ compiler version  : 5.3.1
> --   CXX flags             : -mfpu=neon-vfpv4 -mtune=cortex-a17 -march=armv7ve -mfloat-abi=hard -fvisibility-inlines-hidden -fopenmp -Wnon-virtual-dtor
> --   Build type            : Release
> --   Compile definitions   :
> --   CMAKE_PREFIX_PATH     : /usr/lib/python3/dist-packages
> --   CMAKE_INSTALL_PREFIX  : /root/pytorch_install/pytorch/torch
> --   CMAKE_MODULE_PATH     : /root/pytorch_install/pytorch/cmake/Modules
> --
> --   ONNX version          : 1.4.1
> --   ONNX NAMESPACE        : onnx_torch
> --   ONNX_BUILD_TESTS      : OFF
> --   ONNX_BUILD_BENCHMARKS : OFF
> --   ONNX_USE_LITE_PROTO   : OFF
> --   ONNXIFI_DUMMY_BACKEND : OFF
> --   ONNXIFI_ENABLE_EXT    : OFF
> --
> --   Protobuf compiler     :
> --   Protobuf includes     :
> --   Protobuf libraries    :
> --   BUILD_ONNX_PYTHON     : OFF
> --
> -- ******** Summary ********
> --   CMake version         : 3.5.1
> --   CMake command         : /usr/bin/cmake
> --   System                : Linux
> --   C++ compiler          : /usr/bin/c++
> --   C++ compiler version  : 5.3.1
> --   CXX flags             : -mfpu=neon-vfpv4 -mtune=cortex-a17 -march=armv7ve -mfloat-abi=hard -fvisibility-inlines-hidden -fopenmp -Wnon-virtual-dtor
> --   Build type            : Release
> --   Compile definitions   :
> --   CMAKE_PREFIX_PATH     : /usr/lib/python3/dist-packages
> --   CMAKE_INSTALL_PREFIX  : /root/pytorch_install/pytorch/torch
> --   CMAKE_MODULE_PATH     : /root/pytorch_install/pytorch/cmake/Modules
> --
> --   ONNX version          : 1.4.1
> --   ONNX NAMESPACE        : onnx_torch
> --   ONNX_BUILD_TESTS      : OFF
> --   ONNX_BUILD_BENCHMARKS : OFF
> --   ONNX_USE_LITE_PROTO   : OFF
> --   ONNXIFI_DUMMY_BACKEND : OFF
> --
> --   Protobuf compiler     :
> --   Protobuf includes     :
> --   Protobuf libraries    :
> --   BUILD_ONNX_PYTHON     : OFF
> -- Found gcc >=5 and CUDA <= 7.5, adding workaround C++ flags
> -- Could not find CUDA with FP16 support, compiling without torch.CudaHalfTensor
> -- Removing -DNDEBUG from compile flags
> -- MAGMA not found. Compiling without MAGMA support
> -- No OMAP3 processor on this machine.
> -- No OMAP4 processor on this machine.
> -- Neon found with compiler flag : -mfpu=neon -D__NEON__
> -- Looking for cpuid.h
> -- Looking for cpuid.h - not found
> -- Performing Test NO_GCC_EBX_FPIC_BUG
> -- Performing Test NO_GCC_EBX_FPIC_BUG - Failed
> -- Performing Test C_HAS_AVX_1
> -- Performing Test C_HAS_AVX_1 - Failed
> -- Performing Test C_HAS_AVX_2
> -- Performing Test C_HAS_AVX_2 - Failed
> -- Performing Test C_HAS_AVX_3
> -- Performing Test C_HAS_AVX_3 - Failed
> -- Performing Test C_HAS_AVX2_1
> -- Performing Test C_HAS_AVX2_1 - Failed
> -- Performing Test C_HAS_AVX2_2
> -- Performing Test C_HAS_AVX2_2 - Failed
> -- Performing Test C_HAS_AVX2_3
> -- Performing Test C_HAS_AVX2_3 - Failed
> -- Performing Test CXX_HAS_AVX_1
> -- Performing Test CXX_HAS_AVX_1 - Failed
> -- Performing Test CXX_HAS_AVX_2
> -- Performing Test CXX_HAS_AVX_2 - Failed
> -- Performing Test CXX_HAS_AVX_3
> -- Performing Test CXX_HAS_AVX_3 - Failed
> -- Performing Test CXX_HAS_AVX2_1
> -- Performing Test CXX_HAS_AVX2_1 - Failed
> -- Performing Test CXX_HAS_AVX2_2
> -- Performing Test CXX_HAS_AVX2_2 - Failed
> -- Performing Test CXX_HAS_AVX2_3
> -- Performing Test CXX_HAS_AVX2_3 - Failed
> -- Performing Test HAS_C11_ATOMICS
> -- Performing Test HAS_C11_ATOMICS - Failed
> -- Performing Test HAS_MSC_ATOMICS
> -- Performing Test HAS_MSC_ATOMICS - Failed
> -- Performing Test HAS_GCC_ATOMICS
> -- Performing Test HAS_GCC_ATOMICS - Success
> -- Atomics: using GCC intrinsics
> -- Looking for cheev_
> -- Looking for cheev_ - found
> -- Found a library with LAPACK API (open).
> disabling CUDA because NOT USE_CUDA is set
> -- CuDNN not found. Compiling without CuDNN support
> disabling ROCM because NOT USE_ROCM is set
> -- MIOpen not found. Compiling without MIOpen support
> disabling MKLDNN because USE_MKLDNN is not set
> -- Looking for clock_gettime in rt
> -- Looking for clock_gettime in rt - found
> -- Looking for mmap
> -- Looking for mmap - found
> -- Looking for shm_open
> -- Looking for shm_open - found
> -- Looking for shm_unlink
> -- Looking for shm_unlink - found
> -- Looking for malloc_usable_size
> -- Looking for malloc_usable_size - found
> -- Performing Test C_HAS_THREAD
> -- Performing Test C_HAS_THREAD - Success
> -- GCC 5.3.1: Adding gcc and gcc_s libs to link line
> -- don't use NUMA
> disabling CUDA because USE_CUDA is set false
> -- Check size of long double
> -- Check size of long double - done
> -- Performing Test COMPILER_SUPPORTS_FLOAT128
> -- Performing Test COMPILER_SUPPORTS_FLOAT128 - Failed
> -- Performing Test COMPILER_SUPPORTS_SSE2
> -- Performing Test COMPILER_SUPPORTS_SSE2 - Failed
> -- Performing Test COMPILER_SUPPORTS_SSE4
> -- Performing Test COMPILER_SUPPORTS_SSE4 - Failed
> -- Performing Test COMPILER_SUPPORTS_AVX
> -- Performing Test COMPILER_SUPPORTS_AVX - Failed
> -- Performing Test COMPILER_SUPPORTS_FMA4
> -- Performing Test COMPILER_SUPPORTS_FMA4 - Failed
> -- Performing Test COMPILER_SUPPORTS_AVX2
> -- Performing Test COMPILER_SUPPORTS_AVX2 - Failed
> -- Performing Test COMPILER_SUPPORTS_SVE
> -- Performing Test COMPILER_SUPPORTS_SVE - Failed
> -- Performing Test COMPILER_SUPPORTS_AVX512F
> -- Performing Test COMPILER_SUPPORTS_AVX512F - Failed
> -- Found OpenMP: -fopenmp
> -- Performing Test COMPILER_SUPPORTS_OPENMP
> -- Performing Test COMPILER_SUPPORTS_OPENMP - Success
> -- Performing Test COMPILER_SUPPORTS_WEAK_ALIASES
> -- Performing Test COMPILER_SUPPORTS_WEAK_ALIASES - Success
> -- Performing Test COMPILER_SUPPORTS_BUILTIN_MATH
> -- Performing Test COMPILER_SUPPORTS_BUILTIN_MATH - Success
> -- Configuring build for SLEEF-v3.2
>    Target system: Linux-3.10.0
>    Target processor: armv7l
>    Host system: Linux-3.10.0
>    Host processor: armv7l
>    Detected C compiler: GNU @ /usr/bin/cc
> -- Using option `-Wall -Wno-unused -Wno-attributes -Wno-unused-result -Wno-psabi -ffp-contract=off -fno-math-errno -fno-trapping-math` to compile libsleef
> -- Building shared libs : OFF
> -- MPFR : LIB_MPFR-NOTFOUND
> -- GMP : /usr/lib/arm-linux-gnueabihf/libgmp.so
> -- RUNNING_ON_TRAVIS : 0
> -- COMPILER_SUPPORTS_OPENMP : 1
> -- pytorch is compiling with OpenMP.
> OpenMP CXX_FLAGS: -fopenmp.
> OpenMP libraries: /usr/lib/gcc/arm-linux-gnueabihf/5/libgomp.so;/usr/lib/arm-linux-gnueabihf/libpthread.so.
> -- /usr/bin/c++ /root/pytorch_install/pytorch/torch/abi-check.cpp -o /root/pytorch_install/pytorch/build/abi-check
> -- Determined _GLIBCXX_USE_CXX11_ABI=1
> -- NCCL operators skipped due to no CUDA support
> -- Excluding ideep operators as we are not using ideep
> -- Excluding image processing operators due to no opencv
> -- Excluding video processing operators due to no opencv
> -- MPI operators skipped due to no MPI support
> -- Include Observer library
> -- Caffe2 is compiling with OpenMP.
> OpenMP CXX_FLAGS: -fopenmp.
> OpenMP libraries: /usr/lib/gcc/arm-linux-gnueabihf/5/libgomp.so;/usr/lib/arm-linux-gnueabihf/libpthread.so.
> -- Using lib/python3/dist-packages as python relative installation path
> CMake Warning at CMakeLists.txt:421 (message):
>   Generated cmake files are only fully tested if one builds with system glog,
>   gflags, and protobuf.  Other settings may generate files that are not well
>   tested.
> 
> 
> --
> -- ******** Summary ********
> -- General:
> --   CMake version         : 3.5.1
> --   CMake command         : /usr/bin/cmake
> --   System                : Linux
> --   C++ compiler          : /usr/bin/c++
> --   C++ compiler id       : GNU
> --   C++ compiler version  : 5.3.1
> --   BLAS                  : MKL
> --   CXX flags             : -mfpu=neon-vfpv4 -mtune=cortex-a17 -march=armv7ve -mfloat-abi=hard -fvisibility-inlines-hidden -fopenmp -D_FORCE_INLINES -D_MWAITXINTRIN_H_INCLUDED -D__STRICT_ANSI__ -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -Wno-unused-but-set-variable -Wno-maybe-uninitialized
> --   Build type            : Release
> --   Compile definitions   : ONNX_NAMESPACE=onnx_torch;USE_GCC_ATOMICS=1;HAVE_MMAP=1;_FILE_OFFSET_BITS=64;HAVE_SHM_OPEN=1;HAVE_SHM_UNLINK=1;HAVE_MALLOC_USABLE_SIZE=1
> --   CMAKE_PREFIX_PATH     : /usr/lib/python3/dist-packages
> --   CMAKE_INSTALL_PREFIX  : /root/pytorch_install/pytorch/torch
> --
> --   TORCH_VERSION         : 1.1.0
> --   CAFFE2_VERSION        : 1.1.0
> --   BUILD_ATEN_MOBILE     : OFF
> --   BUILD_ATEN_ONLY       : OFF
> --   BUILD_BINARY          : False
> --   BUILD_CUSTOM_PROTOBUF : ON
> --     Link local protobuf : ON
> --   BUILD_DOCS            : OFF
> --   BUILD_PYTHON          : True
> --     Python version      : 3.5.1
> --     Python executable   : /usr/bin/python3
> --     Pythonlibs version  : 3.5.1+
> --     Python library      : /usr/lib/libpython3.5m.so.1.0
> --     Python includes     : /usr/include/python3.5m
> --     Python site-packages: lib/python3/dist-packages
> --   BUILD_CAFFE2_OPS      : True
> --   BUILD_SHARED_LIBS     : ON
> --   BUILD_TEST            : True
> --   USE_ASAN              : OFF
> --   USE_CUDA              : False
> --   USE_ROCM              : False
> --   USE_EIGEN_FOR_BLAS    : ON
> --   USE_FBGEMM            : OFF
> --   USE_FFMPEG            : False
> --   USE_GFLAGS            : OFF
> --   USE_GLOG              : OFF
> --   USE_LEVELDB           : False
> --   USE_LITE_PROTO        : OFF
> --   USE_LMDB              : False
> --   USE_METAL             : OFF
> --   USE_MKL               : OFF
> --   USE_MKLDNN            : OFF
> --   USE_NCCL              : False
> --   USE_NNPACK            : True
> --   USE_NUMPY             : ON
> --   USE_OBSERVERS         : ON
> --   USE_OPENCL            : OFF
> --   USE_OPENCV            : False
> --   USE_OPENMP            : ON
> --   USE_PROF              : OFF
> --   USE_QNNPACK           : True
> --   USE_REDIS             : OFF
> --   USE_ROCKSDB           : OFF
> --   USE_ZMQ               : OFF
> --   USE_DISTRIBUTED       : False
> --   Public Dependencies  : Threads::Threads
> --   Private Dependencies : qnnpack;nnpack;cpuinfo;fp16;aten_op_header_gen;foxi_loader;rt;gcc_s;gcc;dl
> -- Configuring done
> -- Generating done
> CMake Warning:
>   Manually-specified variables were not used by the project:
> 
>     NCCL_EXTERNAL
>     THD_SO_VERSION
> 
> 
> -- Build files have been written to: /root/pytorch_install/pytorch/build
> Scanning dependencies of target libprotobuf
> Scanning dependencies of target libprotobuf-lite
> [  0%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotobuf-lite.dir/__/src/google/protobuf/arena.cc.o

**Comment by danFromTelAviv** [comment]
I see now that I get the same issue of no pybind that I get when i build caffe2 directly. This may have been the issue before.

**Comment by danFromTelAviv** [comment]
Scanning dependencies of target nnpack
[ 39%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/init.c.o
[ 39%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/convolution-inference.c.o
In file included from /root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/activations.h:25:0,
                 from /root/pytorch_install/pytorch/third_party/NNPACK/src/convolution-inference.c:14:
/root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/arm_neon.h: In function 'vld1q_f32_f16':
/root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/arm_neon.h:46:11: warning: implicit declaration of function 'vcvt_f32_f16' [-Wimplicit-function-declaration]
    return vcvt_f32_f16((float16x4_t) vld1_u16((const uint16_t*) address));
           ^
/root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/arm_neon.h:46:25: error: 'float16x4_t' undeclared (first use in this function)
    return vcvt_f32_f16((float16x4_t) vld1_u16((const uint16_t*) address));
                         ^
/root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/arm_neon.h:46:25: note: each undeclared identifier is reported only once for each function it appears in
/root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/arm_neon.h:46:38: error: expected ')' before 'vld1_u16'
    return vcvt_f32_f16((float16x4_t) vld1_u16((const uint16_t*) address));
                                      ^
/root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/arm_neon.h: In function 'vld1q_f32_f16_aligned':
/root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/arm_neon.h:50:25: error: 'float16x4_t' undeclared (first use in this function)
    return vcvt_f32_f16((float16x4_t)
                         ^
/root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/arm_neon.h:51:5: error: expected ')' before 'vld1_u16'
     vld1_u16((const uint16_t*) __builtin_assume_aligned(address, sizeof(float16x4_t))));
     ^
/root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/arm_neon.h: In function 'vst1q_f16_f32':
/root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/arm_neon.h:55:47: warning: implicit declaration of function 'vcvt_f16_f32' [-Wimplicit-function-declaration]
    vst1_u16((uint16_t*) address, (uint16x4_t) vcvt_f16_f32(vector));
                                               ^
/root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/arm_neon.h:55:4: error: can't convert between vector values of different size
    vst1_u16((uint16_t*) address, (uint16x4_t) vcvt_f16_f32(vector));
    ^
/root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/arm_neon.h: In function 'vst1q_f16_f32_aligned':
/root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/arm_neon.h:60:5: error: can't convert between vector values of different size
     (uint16x4_t) vcvt_f16_f32(vector));
     ^
confu-deps/NNPACK/CMakeFiles/nnpack.dir/build.make:86: recipe for target 'confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/convolution-inference.c.o' failed
make[2]: *** [confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/convolution-inference.c.o] Error 1
CMakeFiles/Makefile2:623: recipe for target 'confu-deps/NNPACK/CMakeFiles/nnpack.dir/all' failed
make[1]: *** [confu-deps/NNPACK/CMakeFiles/nnpack.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/objectivec/objectivec_enum.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/objectivec/objectivec_enum_field.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/objectivec/objectivec_extension.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/objectivec/objectivec_field.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/objectivec/objectivec_file.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/objectivec/objectivec_generator.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/objectivec/objectivec_helpers.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/objectivec/objectivec_map_field.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/objectivec/objectivec_message.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/objectivec/objectivec_message_field.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/objectivec/objectivec_oneof.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/objectivec/objectivec_primitive_field.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/php/php_generator.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/plugin.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/plugin.pb.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/python/python_generator.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/ruby/ruby_generator.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/subprocess.cc.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/zip_writer.cc.o
[ 39%] Linking CXX static library ../../../lib/libprotoc.a
[ 39%] Built target libprotoc
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
Traceback (most recent call last):
  File "setup.py", line 715, in <module>
    build_deps()
  File "setup.py", line 287, in build_deps
    build_dir='build')
  File "/root/pytorch_install/pytorch/tools/build_pytorch_libs.py", line 259, in build_caffe2
    check_call(['make', '-j', str(max_jobs), 'install'], cwd=build_dir, env=my_env)
  File "/usr/lib/python3.5/subprocess.py", line 581, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['make', '-j', '2', 'install']' returned non-zero exit status 2

**Comment by danFromTelAviv** [comment]
[building pytorch arm.txt](https://github.com/pytorch/pytorch/files/2951578/building.pytorch.arm.txt)

**Comment by t-vi** [comment]
So I'm not an ARM expert (following this bug, I bought a Raspberry Pi, though, to improve) but:
```
In function 'vld1q_f32_f16':
/root/pytorch_install/pytorch/third_party/NNPACK/include/nnpack/arm_neon.h:46:11: warning: implicit declaration of function 'vcvt_f32_f16'
```
looks like your arm CPU / compiler does not have neon intrinsics (enabled)?
As far as I know NNPack requires those.

**Comment by danFromTelAviv** [comment]
> So I'm not an ARM expert (following this bug, I bought a Raspberry Pi, though, to improve) but:

:clap: Thank you. 

To the best of my memory all armv7 and 64 bit arm have neon actually. specifically the tinkerboard has -mfpu=neon-vfpv4.
I'll try to remove the cflags and see if it helps maybe.

**Comment by t-vi** [comment]
Maybe the NNPACK instructions can also help: https://github.com/Maratyszcza/NNPACK

**Comment by danFromTelAviv** [comment]
looks like that was the issue. Once again - Keep it simple stupid wins. when I don't add special flags it works. I wonder if it takes full advantage of the architecture though?

Scanning dependencies of target nnpack
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/init.c.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/convolution-inference.c.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/fully-connected-inference.c.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/pooling-output.c.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/relu-output.c.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/softmax-output.c.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/fully-connected-output.c.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/relu-input-gradient.c.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/convolution-input-gradient.c.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/convolution-kernel-gradient.c.o
[ 38%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/java/java_primitive_field_lite.cc.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/convolution-output.c.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/psimd/2d-fourier-8x8.c.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/psimd/2d-fourier-16x16.c.o
[ 38%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/java/java_service.cc.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/2d-winograd-8x8-3x3.c.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/2d-winograd-8x8-3x3-fp16.c.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/blas/h4gemm.c.o
[ 38%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/java/java_shared_code_generator.cc.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/blas/s4gemm.c.o
[ 38%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/blas/c4gemm-conjb.c.o
[ 39%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/blas/s4c2gemm-conjb.c.o
[ 39%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/blas/conv1x1.c.o
[ 39%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/blas/sgemm.c.o
[ 39%] Building ASM object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/blas/h4gemm-aarch32.S.o
[ 39%] Building ASM object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/blas/s4gemm-aarch32.S.o
[ 39%] Building ASM object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/blas/sgemm-aarch32.S.o
[ 39%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/relu.c.o
[ 39%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/psimd/softmax.c.o
[ 39%] Building CXX object third_party/protobuf/cmake/CMakeFiles/libprotoc.dir/__/src/google/protobuf/compiler/java/java_string_field.cc.o
[ 39%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/blas/sdotxf.c.o
[ 39%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/psimd/blas/shdotxf.c.o
[ 39%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/blas/c4gemm.c.o
[ 39%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/blas/s4c2gemm.c.o
[ 39%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/blas/c4gemm-conjb-transc.c.o
[ 39%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/neon/blas/s4c2gemm-conjb-transc.c.o
[ 39%] Linking C static library ../../lib/libnnpack.a

**Comment by danFromTelAviv** [comment]
When it finishes i'll test it for run-time and see if it really comparable to nnpack-darknet / faster than the install without nnpack.

**Comment by danFromTelAviv** [comment]
qnnpack looks a bit suspicious : 
[ 43%] Linking CXX executable ../../../bin/protoc
[ 43%] Built target protoc
Scanning dependencies of target qnnpack
Skipping _th_multinomial Because of Arg: Generator * (Generator*)
Skipping _th_normal Because of Arg: Generator * (Generator*)
Skipping _th_normal Because of Arg: Generator * (Generator*)
Skipping _th_normal Because of Arg: Generator * (Generator*)
Skipping _thnn_rrelu_with_noise_forward Because of Arg: Generator * (Generator*)
Skipping _thnn_conv_transpose2d_backward Because of Arg: std::array<bool,3> (std::array<bool,3>)
Skipping _thnn_conv_transpose3d_backward Because of Arg: std::array<bool,3> (std::array<bool,3>)
Skipping _thnn_conv2d_backward Because of Arg: std::array<bool,3> (std::array<bool,3>)
Skipping _thnn_conv_depthwise2d_backward Because of Arg: std::array<bool,2> (std::array<bool,2>)
Skipping _thnn_conv3d_backward Because of Arg: std::array<bool,3> (std::array<bool,3>)
Skipping _thnn_conv_dilated2d_backward Because of Arg: std::array<bool,3> (std::array<bool,3>)
Skipping _thnn_conv_dilated3d_backward Because of Arg: std::array<bool,3> (std::array<bool,3>)
Skipping _cudnn_rnn_backward Because of Arg: std::array<bool,4> (std::array<bool,4>)
Skipping _cudnn_init_dropout_state because it is a factory method
Skipping _fused_dropout Because of Arg: Generator * (Generator *)
Skipping arange because it is a factory method
Skipping as_strided Because of Arg: c10::optional<int64_t> (int64_t)
[ 43%] Building C object confu-deps/QNNPACK/CMakeFiles/qnnpack.dir/src/init.c.o
Skipping bartlett_window because it is a factory method
Skipping _batch_norm_impl_index_backward Because of Arg: std::array<bool,3> (std::array<bool,3>)
Skipping bernoulli Because of Arg: Generator * (Generator *)
Skipping bernoulli Because of Arg: Generator * (Generator *)
...

**Comment by t-vi** [comment]
So on android/arm, the "not-mobile optimized example style transfer network" went from ridiculous (5 min) to 5 sec or something like that.
The skipping messages are from another thread, probably. - That's about JIT code generation for functions or somesuch.
I don't know anything about QNNPack / PyTorch, unfortunately, that would have been part of my Android effort I had managed to raise commercial interest for it.

**Comment by danFromTelAviv** [comment]
wow ! yeah - from my calculations it should allow tiny yolo v3 to run at around 2 fps. before nnpack it ran a network about 30x smaller at 2 fps. crossing my fingers. 

Thanks again for the unbelievably good support.

**Comment by danFromTelAviv** [comment]
:// it finished installing - everything seemed to go fine. It's not any faster . I am not sure what I did wrong. I did a pip3 uninstall torch before I installed the new version. Is there any way to check if the pytorch I have installed is using nnpack?

**Comment by danFromTelAviv** [comment]
> not-mobile optimized example style transfer network

is [this ](https://heartbeat.fritz.ai/neural-style-transfer-with-pytorch-49e7c1fe3bea) what you were talking about? I want see how fast it runs on my board. Maybe it did really successfully install and is really that slow... who knows

**Comment by danFromTelAviv** [comment]
I tried to reinstall on a fresh image to see if maybe there was some files laying around that are messing with the install. Its still really slow.

**Comment by danFromTelAviv** [comment]
@t-vi can you 

> So on android/arm, the "not-mobile optimized example style transfer network" went from ridiculous (5 min) to 5 sec or something like that.

Can you please refer me to this tutorial so I can try it out and see if get similar results. I want to validate that I installed correctly. ( And i you can tell me how fast it runs on a raspvery pi that would be great ).

**Comment by t-vi** [comment]
It's not so much a tutorial, though I always thought about making one: https://lernapparat.de/pytorch-android/

**Comment by danFromTelAviv** [comment]
I don't see the network architecture there? It's vaguely described. Can you refer me to the model?

**Comment by danFromTelAviv** [comment]
@t-vi  I want to give this one more go. It doesn't make sense to me that pytorch with and without nnpack perform the same on my board. Especially since I did see a boost in performance when I used nnpack + darknet. 
I tried to install both on separate fresh boards to avoid issues ( just btw ). 
Can you please review by logs and see if you see anything strange or try to see if on the raspberry pi you get a boost in speed? 
Thanks,
Dan

**Comment by elepherai** [comment]
@danFromTelAviv  Hi Dan, do you have any progress on this experiment? w/o nnpack.

**Comment by danFromTelAviv** [comment]
@elepherai no. i never got it to run faster unfortunately. 
i did 2 alternative things. 

the first is i create a library of basic operations in cython. i haven't really been maintaining it but its a good starting point to show you how and what. i actually did create a working stacked qrnn network but it's proprietary and i can't share it. 

the second thing i did was ask keras to create an np backend. they roled with me and i helped a bit with the coding. it's part of keras in production now, but it's not very fast at all...

**Comment by elepherai** [comment]
@danFromTelAviv ok, thank you for sharing your alternative scenarios.

**Comment by t-vi** [comment]
From what I can see, building with NNPack and QNNPack works out of the box on both raspbian and arm64 (Raspbian 64 bit kernel + Debian arm64).
So I think it might be time to close the bug. (And maybe have a new more specific one for other things.)

**Comment by danFromTelAviv** [comment]
@t-vi 
In theory it did build with nnpack. In practice it wasn't any faster than building without nnpack. This is strange since other libraries were faster with nnpack bindings. 
I left my previous job where this was relevant for me but I really do think it's worth investigating farther. The speed doesn't add up.

**Comment by t-vi** [comment]
I know that there there are people interested in benchmarking and they'd appreciate any reasonably concrete hints. My usual method is to build a small C++ extension wrapping the would-be faster function and run it.
So one thing that surfaced is that we currently don't use the inference part.
